docs/Threading.md

```diff
@@ -0,0 +1,167 @@
+The SGX-LKL threading implementation
+====================================
+
+SGX-LKL has four kinds of threads: ethreads, lthreads, Linux tasks, and pthreads.
+This document explains how they are implemented and how they relate.
+
+Ethreads are the lowest-level threads are ethreads.
+These are roughly analogous to VCPUs in conventional VMs.
+In SGX, each ethread corresponds to a thread-control structure (TCS), allocated by OpenEnclave.
+
+The lthreads layer provides a cooperative threading library on top of ethreads.
+Each ethread runs an instance of the lthread scheduler, which pulls lthreads from a multi-producer, multi-consumer queue (MPMCQ) to run.
+These threads run until they explicitly yield.
+
+The Linux kernel (LKL) manages its own task abstraction.
+Each Linux task is associated with an lthread.
+The LKL documentation refers to 'host threads', these are lthreads in our use.
+
+The userspace libc (musl) provides a POSIX threads (pthreads) implementation that sits atop Linux's tasks.
+Each pthread is run by the lthread scheduler but may be caused to block by LKL.
+
+Ethreads
+========
+
+Ethreads are created by the host environment calling one of two ecalls, either [`sgxlkl_enclave_init`](https://github.com/lsds/sgx-lkl/blob/24467b08346cd7384eb93f845dec896a1d429711/src/enclave/enclave_oe.c#L395) (for the first ethread) or [`sgxlkl_ethread_init`](https://github.com/lsds/sgx-lkl/blob/24467b08346cd7384eb93f845dec896a1d429711/src/enclave/enclave_oe.c) (for subsequent ethreads).
+Each of these threads calls `lthread_run` to start an instance of the lthread scheduler.
+
+The ethreads do not use the FS segment on x86-64.
+The GS segment stores a pointer to the Open Enclave state and the lthread scheduler.
+The lthread context switch routine switches the FS segment register and makes this available for higher layers in the threading stack
```

Comment:
a link to the context switch routine would be helpful here.

Reply:
```suggestion
The [lthread context switch](https://github.com/lsds/sgx-lkl/blob/796b346d4762a93be7bfc2ba5ce97b9ab840a4bd/src/sched/lthread.c#L466) [switches the FS segment register](https://github.com/lsds/sgx-lkl/blob/796b346d4762a93be7bfc2ba5ce97b9ab840a4bd/src/sched/lthread.c#L411) and makes this available for higher layers in the threading stack.
```